File: algo/toolbox/fibonacci_huge.cpp
#include <iostream>
#include <vector>


long long get_fibonacci_huge_naive(long long n, long long m) {
    if (n <= 1)
        return n;

    long long previous = 0;
    long long current  = 1;

    for (long long i = 0; i < n - 1; ++i) {
        long long tmp_previous = previous;
        previous = current;
        current = tmp_previous + current;
    }

    return current % m;
}


long long get_fibonacci_huge_fast(long long n, long long m) {
    if (n <= 1)
        return n;

    long long pizano_row[m * m];
    pizano_row[0] = 0;
    pizano_row[1] = 1;

    long long i = 2;
    while (true) {
       pizano_row[i] = (pizano_row[i - 1] + pizano_row[i - 2]) % m;

       if (pizano_row[i] == 1 && pizano_row[i - 1] == 0 && i > 2)
           break;
        i++;
    }

    long long pizano_len = i - 1;

    return pizano_row[n % pizano_len];
}

/*
int main() {
    long long n, m;
    std::cin >> n >> m;
    std::cout << get_fibonacci_huge_fast(n, m) << '\n';

    return 0;
}
*/

File: algo/toolbox/fibonacci_huge.py
# Uses python3
import sys


def get_fibonacci_huge_naive(n, m):
    if n <= 1:
        return n

    previous = 0
    current = 1

    for _ in range(n - 1):
        previous, current = current, previous + current

    return current % m


def get_fibonacci_huge_fast(n, m):
    if n <= 1:
        return n

    pizano_row = list()
    pizano_row.append(0)
    pizano_row.append(1)

    i = 2
    while True:
        pizano_row.append((pizano_row[i - 1] + pizano_row[i - 2]) % m)

        if pizano_row[i] == 1 and pizano_row[i - 1] == 0 and i != 2:
            break
        i += 1

    return pizano_row[n % (i - 1)]


if __name__ == '__main__':
    n, m = map(int, input().split())
    print(get_fibonacci_huge_fast(n, m))
